_posts/2017-09-05-Browser Security Features.md

```diff
@@ -0,0 +1,290 @@
+
+Afterwards, you can register your domain at [https://hstspreload.org/](https://hstspreload.org/){: target="_blank" }
+
+### Are there any risks?
+Activating HSTS does offer some risks:
+* If you include the `includesubdomains` directive, you tell the browser that *all* subdomains need to be retrieved over HTTPS.
+  If your internal applications are on a subdomain (e.g. internal.example.com), you'll block access to those that haven't enabled TLS yet.
+* Adding the `preload` directive is even more dangerous: because this tells browser makes to hardcode your HSTS settings.
+  If you've made a mistake in the setup, it can take a long time be removed from the list. 
+  Since this list is *in* the browser, you'll affect both your existing and your new users.
+These risks can be mitigated through extensive testing and conservative settings. 
+Start with a short max-age and slowly increase its length, don't include subdomains if you're not 100% sure that *all* subdomains need to be included
+and perhaps most importantly, **don't** activate preload unless you're 100% sure that everything works as intended.
+
+### Should I activate HSTS?
+For those (subdomains) where TLS is enabled, you should start rolling out HSTS (while keeping in mind the warnings above).
+
+## Public key pinning
+Alright, now you've secured your site with HTTPS, and you've made sure your users can't fall victim to a Man in the Middle attack.
+Or have you?
+It's true that HSTS will make sure that the user only connects using a secured connection, but that doesn't mean the HTTPS connection is actually made to **your** server.
+Over the last couple of years, there were several incidents where malicious actors were able to generate valid certificates for domains they didn't control.
+When this happens, your users will think they're safe (as their browser shows the green padlock), but the attacker can still manipulate your content.
+To protect yourself against this, there's a mechanism called "HTTP Public Key Pinning" (HPKP).
+With HPKP you "pin" the public key of your TLS certificate to the browser.
+In the future, that browser will compare the public key that's actually used for the TLS connection with the pinned one and, if they don't match, refuse the connection altogether.
+![HPPK error](/img/2017-security-features/hpkp-error.png){: .image height="245px" }
+
+An HPKP header looks like this:
+
+{% highlight http %}
+public-key-pins:pin-sha256="YLh1...uihg=";pin-sha256="9dNi...Dwg=";pin-sha256="Vjs...eWys=";max-age=2592000;includeSubdomains;report-uri="report-uri"
+{% endhighlight %}{: style="scroll: auto" }
+
+The `max-age` directive tells the browser for how long these pins are valid. 
+You can use `report-uri` to get a report when an invalid certificate is used.
+`includesubdomains` makes sure that the policy also applies to your subdomains.
+Finally, there are the actual pins.
+You need to pin at least 2 fingerprints: 1 that should be active at the moment and 1 that isn't.
```

Comment:
1 -> one

Reply:
I try to be consistent: numbers are written numerically, rather than in full.